### PR TITLE
DO NOT SUBMIT: update package:build interfaces to support reading/writing as bytes

### PR DESCRIPTION
**Pull request by jakemac53**
Preview of a number of changes to the apis, primarily to support reading/writing as bytes,
as this is going to inevitably be a required feature.

### New Features
- The `AssetWriter` class now has a
  `Future writeAsBytes(AssetId id, List<int> bytes)` method.
- The `AssetReader` class now has a `Future<List<int>> readAsBytes(AssetId id)`
  method.

### Breaking Changes
- The `Asset` class has been removed entirely.
- The `AssetWriter#writeAsString` signature has changed to
  `Future writeAsString(AssetId id, String contents, {Encoding encoding})`.
- The type of the `AssetWriterSpy#assetsWritten` getter has changed from an
  `Iterable<Asset>` to an `Iterable<AssetId>`.
- `BuildStep#input` has been changed to `BuildStep#inputId`, and its type has
  changed from `Asset` to `AssetId`. This means you must now use
  `BuildStep#readAsString` or `BuildStep#readAsBytes` to read the primary input,
  instead of it already being read in for you.

**Note**: The changes to `AssetReader` and `AssetWriter` also affect `BuildStep`
and other classes that implement those interfaces.

### New Features
- The `testBuilder` method now accepts `List<int>` values for both
  `sourceAssets` and `outputs`.

### Breaking Changes
- The `testBuilder` method now requires a `RecordingAssetWriter` instead of
  just an `AssetWriter` for the `writer` parameter.
- Deleted the `makeAsset` and `makeAssets` methods. There is no more `Asset`
  class so these don't really have any value any more.
- The signature of `addAssets` has changed to
  `void addAssets(Map<AssetId, dynamic> assets, InMemoryAssetWriter writer)`.
  Values of the map may be either `String` or `List<int>`.
- `InMemoryAssetReader#assets` and `InMemoryAssetWriter#assets` have changed to
  a type of `Map<AssetId, DatedValue>` from a type of
  `Map<AssetId, DatedString>`. `DatedValue` has both a `stringValue` and
  `bytesValue` getter.
- `InMemoryAssetReader` and `InMemoryAssetWriter` have been updated to implement
  the new `AssetReader` and `AssetWriter` interfaces (see the `build` package
  CHANGELOG for more details).
- `InMemoryAssetReader#cacheAsset` has been changed to two separate methods,
  `void cacheStringAsset(AssetId id, String contents)` and
  `void cacheBytesAsset(AssetId id, List<int> bytes)`.
- The `equalsAsset` matcher has been removed, since there is no more `Asset`
  class.

### CODE REVIEW

**Comment by jakemac53** [comment]
Resolves https://github.com/dart-lang/build/issues/165

**Comment by jakemac53** [comment]
Fwiw, I also considered changing `writeAsString` and `writeAsBytes` to not take an `Asset`, but just and `AssetId` and a value (`String` or `List<int>`), and then possibly adding a `writeAsset` method which just takes an `Asset` (of any type). This would be a larger breaking change though.

**Comment by jakemac53** [comment]
Updated based on offline talks, removed the `Asset` class entirely to simplify/unify things.

File: build/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,29 @@
+## 0.7.0-dev
+
+A number of changes to the apis, primarily to support reading/writing as bytes,
+as this is going to inevitably be a required feature. This will hopefully be the
+last breaking change before the `1.0` release, but it is a fairly large one.
```
Comment (natebosch) [inline build/CHANGELOG.md]: [nit] remove this comment since we have another breaking change planned

File: build/lib/src/asset/writer.dart [hunk 2]
```diff
@@ -4,24 +4,33 @@
 import 'dart:async';
 import 'dart:convert';
 
-import 'asset.dart';
+import 'id.dart';
 
+/// Abstract class that can write [Asset]s.
```
Comment (natebosch) [inline build/lib/src/asset/writer.dart]: [nit] This comment isn't adding much information over the signature. We get a link to [Asset] but I don't think it's worth the comment

File: build/lib/src/asset/writer.dart [hunk 3]
```diff
@@ -4,24 +4,33 @@
...
+
+  Future writeAsString(AssetId id, String contents, {Encoding encoding: UTF8});
 }
 
-/// An [AssetWriter] which tracks all assets written during it's lifetime.
+/// An [AssetWriter] which tracks all [AssetId]s written during it's lifetime.
 class AssetWriterSpy implements AssetWriter {
   final AssetWriter _delegate;
-  final List<Asset> _assetsWritten = [];
+  final List<AssetId> _assetsWritten = [];
```
Comment (natebosch) [inline build/lib/src/asset/writer.dart]: while we're in this file... this can be a Set

File: build/lib/src/builder/build_step.dart [hunk 4]
```diff
@@ -26,19 +25,29 @@ abstract class BuildStep implements AssetReader, AssetWriter {
   @override
   Future<bool> hasInput(AssetId id);
 
+  /// Reads an [Asset] by [id] as a [List<int>].
+  @override
+  Future<List<int>> readAsBytes(AssetId id);
```
Comment (natebosch) [inline build/lib/src/builder/build_step.dart]: I should have noticed this earlier... but repeating these signatures here isn't adding any value. The `implements` statement is all we need here. You can move these doc comments to the other classes if they aren't already there.

File: build/lib/src/builder/build_step.dart [hunk 5]
```diff
@@ -26,19 +25,29 @@ abstract class BuildStep implements AssetReader, AssetWriter {
...
+
   /// Reads an [Asset] by [id] as a [String] using [encoding].
   @override
   Future<String> readAsString(AssetId id, {Encoding encoding: UTF8});
 
-  /// Outputs an [Asset] using the current [AssetWriter], and adds [asset] to
-  /// [outputs].
+  /// Outputs an [BytesAsset] using the current [AssetWriter].
+  ///
+  /// Most `Builder` implementations should not need to `await` this Future
```
Comment (natebosch) [inline build/lib/src/builder/build_step.dart]: I suppose this bit of information could be worth adding. Feel free to keep the method signature and doc comment here

File: build/lib/src/builder/build_step_impl.dart [hunk 6]
```diff
@@ -43,11 +42,11 @@ class BuildStepImpl implements ManagedBuildStep {
...
 
-  BuildStepImpl(Asset input, Iterable<AssetId> expectedOutputs, this._reader,
-      this._writer, this._rootPackage, this._resolvers,
+  BuildStepImpl(AssetId inputId, Iterable<AssetId> expectedOutputs,
+      this._reader, this._writer, this._rootPackage, this._resolvers,
       {Logger logger})
-      : this.input = input,
-        this.logger = logger ?? new Logger('${input.id}') {
+      : this.inputId = inputId,
+        this.logger = logger ?? new Logger('${inputId}') {
```
Comment (natebosch) [inline build/lib/src/builder/build_step_impl.dart]: [nit] remove the {} from the string interpolation

File: build/pubspec.yaml [hunk 7]
```diff
@@ -18,3 +18,9 @@ dev_dependencies:
   glob: ^1.1.0
   test: ^0.12.0
   transformer_test: ^0.2.0
+
+dependency_overrides:
```
Comment (natebosch) [inline build/pubspec.yaml]: drop before merging

File: build/CHANGELOG.md (hunk 1 above)
Comment (jakemac53) [inline build/CHANGELOG.md]: I am going to leave it in for now, as I think we should wait to release until we have all breaking changes done. I set up a milestone for this (0.7.0) to track other changes, so we can continue to add to the changelog for this version.

File: build/lib/src/asset/writer.dart (hunk 2 above)
Comment (jakemac53) [inline build/lib/src/asset/writer.dart]: and `Asset` isn't even a thing any more anyways haha, done

File: build/lib/src/asset/writer.dart (hunk 3 above)
Comment (jakemac53) [inline build/lib/src/asset/writer.dart]: done

File: build/lib/src/builder/build_step.dart (hunk 4 above)
Comment (jakemac53) [inline build/lib/src/builder/build_step.dart]: done, I left the `buildStep` specific comments here

File: build/lib/src/builder/build_step.dart (hunk 5 above)
Comment (jakemac53) [inline build/lib/src/builder/build_step.dart]: yep added this paragraph but removed the general comment

File: build/lib/src/builder/build_step_impl.dart (hunk 6 above)
Comment (jakemac53) [inline build/lib/src/builder/build_step_impl.dart]: done

File: build/pubspec.yaml (hunk 7 above)
Comment (jakemac53) [inline build/pubspec.yaml]: Actually, I was planning on leaving these in until the next release. Otherwise no tests will pass etc.

Comment (jakemac53) [inline build/pubspec.yaml]: I added an additional issue to the milestone to remove these

**Comment by jakemac53** [comment]
Ok, I pushed a commit that updates `build_test` now as well, PTAL.

This is necessary to get tests and things passing, I will update the other packages separately.

File: build_test/lib/src/assets.dart [hunk 8]
```diff
@@ -14,22 +14,15 @@ AssetId makeAssetId([String assetIdString]) {
...
-    var asset = makeAsset(idString, content);
-    assets[asset.id] = asset;
+void addAssets(Map<AssetId, dynamic> assets, InMemoryAssetWriter writer) {
+  assets.forEach((id, value) {
+    if (value is String) {
+      writer.assets[id] = new DatedString(value);
+    } else if (value is List<int>) {
+      writer.assets[id] = new DatedBytes(value);
+    } else {
+      throw 'asset values must be of type `String` or `List<int>`, got '
```
Comment (natebosch) [inline build_test/lib/src/assets.dart]: throw new ArgumentError?

**Pull request merged**